File: app/constructs/KubernetesNodePortService.cs
using System;
using System.IO;
using System.Collections.Generic;
using Constructs;
using HashiCorp.Cdktf;
using HashiCorp.Cdktf.Providers.Kubernetes.Service;


namespace MyCompany.MyApp
{

    class KubernetesNodePortServiceConfig
    {
        public int Port { get; set; }
        public string App { get; set; }
        public string Component { get; set; }
        public string Environment { get; set; }
    }

    class KubernetesNodePortService : Construct
    {
        public Service Resource { get; set; }

        public KubernetesNodePortService(Construct scope, string id, KubernetesNodePortServiceConfig config) : base(scope, id)
        {

            this.Resource = new Service(this, id, new ServiceConfig
            {
                Metadata = new ServiceMetadata
                {
                    Name = $"{config.App}-{config.Component}-{config.Environment}",
                },
                Spec = new ServiceSpec
                {
                    Type = "NodePort",
                    Selector = new Dictionary<string, string> {
                        { "app", config.App },
                        { "component", config.Component },
                        { "environment", config.Environment }
                    },
                    Port = new ServiceSpecPort[]
                    {
                        new ServiceSpecPort
                        {
                            Port = 80,
                            TargetPort = "80",
                            NodePort = config.Port,
                            Protocol = "TCP"
                        }
                    }
                }
            });
        }
    }
}

File: cdktf-examples/java/src/main/java/com/mycompany/constructs/KubernetesNodePortService.java
package com.mycompany.constructs;

import java.util.List;
import java.util.Map;

import org.jetbrains.annotations.NotNull;

import com.hashicorp.cdktf.providers.kubernetes.service.Service;
import com.hashicorp.cdktf.providers.kubernetes.service.ServiceConfig;
import com.hashicorp.cdktf.providers.kubernetes.service.ServiceMetadata;
import com.hashicorp.cdktf.providers.kubernetes.service.ServiceSpec;
import com.hashicorp.cdktf.providers.kubernetes.service.ServiceSpecPort;

import software.constructs.Construct;

public class KubernetesNodePortService extends Construct {
    public final Service service;

    public KubernetesNodePortService(@NotNull Construct scope, @NotNull String id,
            @NotNull KubernetesNodePortServiceConfig config) {
        super(scope, id);

        this.service = new Service(this, id, ServiceConfig.builder()
                .metadata(ServiceMetadata.builder()
                        .name(String.format("%s-%s-%s", config.app(),
                                config.component(), config.environment()))
                        .build())
                .spec(ServiceSpec.builder()
                        .type("NodePort")
                        .port(List.of(
                                ServiceSpecPort.builder()
                                        .port(80)
                                        .targetPort("80")
                                        .nodePort(config.port())
                                        .protocol("TCP")
                                        .build()))
                        .selector(Map.of(
                                "app", config.app(),
                                "component", config.component(),
                                "environment", config.environment()))
                        .build())
                .build());
    }

}
